// @lowgular/playgular/src/lib/routing/visit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.goToAndWait = exports.goTo = exports.expectAutomaticRedirect = exports.waitForURL = exports.expectUrlRegexp = exports.expectUrl = void 0;
const tslib_1 = require("tslib");
const test_1 = require("@playwright/test");
const const_1 = require("../const");
const matchers_1 = require("../matchers");
const expectUrl = (page, pageUrl) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    yield page.waitForURL(pageUrl, const_1.WAIT_CONFIG);
    (0, test_1.expect)(page.url()).toEqual(pageUrl);
});
exports.expectUrl = expectUrl;
const expectUrlRegexp = (page, pageUrl, baseUrl) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    const actualPageUrl = page.url();
    if (baseUrl) {
        (0, test_1.expect)(actualPageUrl).toMatch(new RegExp('^' + (0, matchers_1.escapeRegExp)(baseUrl), 'i'));
    }
    (0, test_1.expect)(page.url()).toMatch(pageUrl);
});
exports.expectUrlRegexp = expectUrlRegexp;
const waitForURL = (page, url) => page.waitForURL((0, matchers_1.coerceRegexp)(url), const_1.WAIT_CONFIG);
exports.waitForURL = waitForURL;
const expectAutomaticRedirect = (page, destinationUrl, redirectUrl) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    yield page.goto(destinationUrl);
    yield page.waitForURL(redirectUrl, const_1.WAIT_CONFIG);
});
exports.expectAutomaticRedirect = expectAutomaticRedirect;
const goTo = (page, destinationUrl) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    yield page.goto(destinationUrl);
});
exports.goTo = goTo;
const goToAndWait = (page, destinationUrl) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    console.log(destinationUrl);
    yield page.goto(destinationUrl);
    yield page.waitForURL(destinationUrl, const_1.WAIT_CONFIG);
});
exports.goToAndWait = goToAndWait;
